File: backend/routers/notifications.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from backend.auth import get_current_user
from backend.database import get_db
from backend.models import User
from backend.models.notification import FCMDevice, Notification, NotificationPreference
from backend.schemas.notification import (
    FCMDeviceCreate,
    FCMDeviceRead,
    NotificationPreferenceRead,
    NotificationPreferenceUpdate,
    NotificationRead,
    SendNotificationRequest,
)
from backend.services.notification_service import (
    get_or_create_preferences,
    send_notification,
)

router = APIRouter()


# ---------------------------------------------------------------------------
# Notifications CRUD
# ---------------------------------------------------------------------------


@router.get("/", response_model=list[NotificationRead])
def list_notifications(
    unread_only: bool = False,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """List notifications for the authenticated user."""
    query = db.query(Notification).filter(Notification.user_id == current_user.id)
    if unread_only:
        query = query.filter(Notification.is_read == False)  # noqa: E712
    return query.order_by(Notification.created_at.desc()).all()


@router.get("/unread-count")
def unread_count(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Return the number of unread notifications."""
    count = (
        db.query(Notification)
        .filter(
            Notification.user_id == current_user.id,
            Notification.is_read == False,  # noqa: E712
        )
        .count()
    )
    return {"unread_count": count}


@router.get("/{notification_id}", response_model=NotificationRead)
def get_notification(
    notification_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    notification = (
        db.query(Notification)
        .filter(
            Notification.id == notification_id,
            Notification.user_id == current_user.id,
        )
        .first()
    )
    if not notification:
        raise HTTPException(status_code=404, detail="Notification not found")
    return notification


@router.patch("/{notification_id}/read", response_model=NotificationRead)
def mark_as_read(
    notification_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Mark a single notification as read."""
    notification = (
        db.query(Notification)
        .filter(
            Notification.id == notification_id,
            Notification.user_id == current_user.id,
        )
        .first()
    )
    if not notification:
        raise HTTPException(status_code=404, detail="Notification not found")
    notification.is_read = True
    db.commit()
    db.refresh(notification)
    return notification


@router.post("/mark-all-read")
def mark_all_read(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Mark all notifications as read for the authenticated user."""
    db.query(Notification).filter(
        Notification.user_id == current_user.id,
        Notification.is_read == False,  # noqa: E712
    ).update({"is_read": True})
    db.commit()
    return {"detail": "All notifications marked as read"}


@router.delete("/{notification_id}", status_code=204)
def delete_notification(
    notification_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    notification = (
        db.query(Notification)
        .filter(
            Notification.id == notification_id,
            Notification.user_id == current_user.id,
        )
        .first()
    )
    if not notification:
        raise HTTPException(status_code=404, detail="Notification not found")
    db.delete(notification)
    db.commit()
    return None


# ---------------------------------------------------------------------------
# Notification preferences
# ---------------------------------------------------------------------------


@router.get("/preferences/me", response_model=NotificationPreferenceRead)
def get_preferences(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Get notification preferences for the authenticated user."""
    prefs = get_or_create_preferences(db, current_user.id)
    return prefs


@router.put("/preferences/me", response_model=NotificationPreferenceRead)
def update_preferences(
    data: NotificationPreferenceUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Update notification preferences for the authenticated user."""
    prefs = get_or_create_preferences(db, current_user.id)
    for key, value in data.model_dump(exclude_unset=True).items():
        setattr(prefs, key, value)
    db.commit()
    db.refresh(prefs)
    return prefs


# ---------------------------------------------------------------------------
# FCM device tokens
# ---------------------------------------------------------------------------


@router.get("/devices/me", response_model=list[FCMDeviceRead])
def list_devices(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """List registered FCM device tokens for the authenticated user."""
    return db.query(FCMDevice).filter(FCMDevice.user_id == current_user.id).all()


@router.post("/devices", response_model=FCMDeviceRead)
def register_device(
    data: FCMDeviceCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Register a new FCM device token. Idempotent – if the token already
    exists for this user the existing record is returned."""
    existing = (
        db.query(FCMDevice)
        .filter(FCMDevice.token == data.token, FCMDevice.user_id == current_user.id)
        .first()
    )
    if existing:
        return existing

    device = FCMDevice(
        user_id=current_user.id,
        token=data.token,
        device_name=data.device_name,
    )
    db.add(device)
    db.commit()
    db.refresh(device)
    return device


@router.delete("/devices/{device_id}", status_code=204)
def remove_device(
    device_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Remove an FCM device token."""
    device = (
        db.query(FCMDevice)
        .filter(FCMDevice.id == device_id, FCMDevice.user_id == current_user.id)
        .first()
    )
    if not device:
        raise HTTPException(status_code=404, detail="Device not found")
    db.delete(device)
    db.commit()
    return None


# ---------------------------------------------------------------------------
# Send notification (for internal / admin use)
# ---------------------------------------------------------------------------


@router.post("/send", response_model=NotificationRead)
def send(
    data: SendNotificationRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Send a notification to a specific user via all enabled channels."""
    target_user = db.query(User).filter(User.id == data.user_id).first()
    if not target_user:
        raise HTTPException(status_code=404, detail="Target user not found")

    notification = send_notification(
        db=db,
        user_id=data.user_id,
        title=data.title,
        body=data.body,
        notification_type=data.notification_type,
        data=data.data,
    )
    return notification

File: backend/src/routes/notifications.js
import { Router } from 'express';
import db from '../database.js';
import { authenticate } from '../auth.js';
import { getOrCreatePreferences, sendNotification, testPushToUser } from '../services/notificationService.js';

const router = Router();

// All notification routes require authentication
router.use(authenticate);

// ---------------------------------------------------------------------------
// Notifications CRUD
// ---------------------------------------------------------------------------

router.get('/', async (req, res) => {
  try {
    const tenantId = req.user.tenant_id;
    let query = db('notifications').where({ user_id: req.user.id, tenant_id: tenantId });
    if (req.query.unread_only === 'true') {
      query = query.where({ is_read: false });
    }
    const notifications = await query.orderBy('created_at', 'desc');
    res.json(notifications);
  } catch (err) {
    console.error('List notifications error:', err);
    res.status(500).json({ detail: 'Internal server error' });
  }
});

router.get('/unread-count', async (req, res) => {
  try {
    const tenantId = req.user.tenant_id;
    const [{ count }] = await db('notifications')
      .where({ user_id: req.user.id, tenant_id: tenantId, is_read: false })
      .count('* as count');
    res.json({ unread_count: count });
  } catch (err) {
    console.error('Unread count error:', err);
    res.status(500).json({ detail: 'Internal server error' });
  }
});

router.get('/preferences/me', async (req, res) => {
  try {
    const prefs = await getOrCreatePreferences(req.user.id);
    res.json(prefs);
  } catch (err) {
    console.error('Get preferences error:', err);
    res.status(500).json({ detail: 'Internal server error' });
  }
});

router.put('/preferences/me', async (req, res) => {
  try {
    const prefs = await getOrCreatePreferences(req.user.id);

    const updates = {};
    for (const key of ['push_enabled', 'email_enabled', 'task_assigned', 'task_updated', 'shift_reminder', 'project_updated']) {
      if (req.body[key] !== undefined) updates[key] = req.body[key];
    }

    if (Object.keys(updates).length > 0) {
      await db('notification_preferences').where({ id: prefs.id }).update(updates);
    }

    const updated = await db('notification_preferences').where({ id: prefs.id }).first();
    res.json(updated);
  } catch (err) {
    console.error('Update preferences error:', err);
    res.status(500).json({ detail: 'Internal server error' });
  }
});

router.get('/devices/me', async (req, res) => {
  try {
    const devices = await db('fcm_devices').where({ user_id: req.user.id });
    res.json(devices);
  } catch (err) {
    console.error('List devices error:', err);
    res.status(500).json({ detail: 'Internal server error' });
  }
});

router.post('/devices', async (req, res) => {
  try {
    const { token, device_name } = req.body;
    if (!token) return res.status(422).json({ detail: 'token is required' });

    // Check if this token already exists (for any user)
    const existing = await db('fcm_devices').where({ token }).first();

    if (existing) {
      // Token already registered — reassign to current user if needed
      if (existing.user_id !== req.user.id || existing.device_name !== (device_name || null)) {
        await db('fcm_devices').where({ id: existing.id }).update({
          user_id: req.user.id,
          device_name: device_name || null,
        });
        const updated = await db('fcm_devices').where({ id: existing.id }).first();
        return res.json(updated);
      }
      return res.json(existing);
    }

    try {
      const [id] = await db('fcm_devices').insert({
        user_id: req.user.id,
        token,
        device_name: device_name || null,
      });
      const device = await db('fcm_devices').where({ id }).first();
      return res.json(device);
    } catch (insertErr) {
      // Handle race condition: token was inserted between our check and insert
      if (insertErr.code === 'SQLITE_CONSTRAINT_UNIQUE' || insertErr.message?.includes('UNIQUE constraint')) {
        const conflict = await db('fcm_devices').where({ token }).first();
        if (conflict) {
          await db('fcm_devices').where({ id: conflict.id }).update({
            user_id: req.user.id,
            device_name: device_name || null,
          });
          const updated = await db('fcm_devices').where({ id: conflict.id }).first();
          return res.json(updated);
        }
      }
      throw insertErr;
    }
  } catch (err) {
    console.error('Register device error:', err);
    res.status(500).json({ detail: 'Internal server error' });
  }
});

router.delete('/devices/:deviceId', async (req, res) => {
  try {
    const device = await db('fcm_devices')
      .where({ id: req.params.deviceId, user_id: req.user.id })
      .first();
    if (!device) return res.status(404).json({ detail: 'Device not found' });

    await db('fcm_devices').where({ id: req.params.deviceId }).del();
    res.status(204).send();
  } catch (err) {
    console.error('Remove device error:', err);
    res.status(500).json({ detail: 'Internal server error' });
  }
});

// Send a test push notification to the current user (bypasses preferences)
router.post('/test-push', async (req, res) => {
  try {
    const result = await testPushToUser(req.user.id);
    const status = result.success ? 200 : 422;
    res.status(status).json(result);
  } catch (err) {
    console.error('Test push error:', err);
    res.status(500).json({ success: false, error: 'Internal server error', devices: [] });
  }
});

router.post('/send', async (req, res) => {
  try {
    const tenantId = req.user.tenant_id;
    const { user_id, title, body, notification_type, data } = req.body;
    if (!user_id || !title || !body) {
      return res.status(422).json({ detail: 'user_id, title, and body are required' });
    }

    const targetUser = await db('users').where({ id: user_id, tenant_id: tenantId }).first();
    if (!targetUser) return res.status(404).json({ detail: 'Target user not found' });

    const notification = await sendNotification({
      userId: user_id,
      tenantId: targetUser.tenant_id,
      title,
      body,
      notificationType: notification_type || 'general',
      data,
    });
    res.json(notification);
  } catch (err) {
    console.error('Send notification error:', err);
    res.status(500).json({ detail: 'Internal server error' });
  }
});

router.post('/mark-all-read', async (req, res) => {
  try {
    const tenantId = req.user.tenant_id;
    await db('notifications')
      .where({ user_id: req.user.id, tenant_id: tenantId, is_read: false })
      .update({ is_read: true });
    res.json({ detail: 'All notifications marked as read' });
  } catch (err) {
    console.error('Mark all read error:', err);
    res.status(500).json({ detail: 'Internal server error' });
  }
});

// Parameterized routes MUST come last to avoid matching named routes
router.get('/:notificationId', async (req, res) => {
  try {
    const tenantId = req.user.tenant_id;
    const notification = await db('notifications')
      .where({ id: req.params.notificationId, user_id: req.user.id, tenant_id: tenantId })
      .first();
    if (!notification) return res.status(404).json({ detail: 'Notification not found' });
    res.json(notification);
  } catch (err) {
    console.error('Get notification error:', err);
    res.status(500).json({ detail: 'Internal server error' });
  }
});

router.patch('/:notificationId/read', async (req, res) => {
  try {
    const tenantId = req.user.tenant_id;
    const notification = await db('notifications')
      .where({ id: req.params.notificationId, user_id: req.user.id, tenant_id: tenantId })
      .first();
    if (!notification) return res.status(404).json({ detail: 'Notification not found' });

    await db('notifications').where({ id: req.params.notificationId, tenant_id: tenantId }).update({ is_read: true });

    const updated = await db('notifications').where({ id: req.params.notificationId }).first();
    res.json(updated);
  } catch (err) {
    console.error('Mark as read error:', err);
    res.status(500).json({ detail: 'Internal server error' });
  }
});

router.delete('/:notificationId', async (req, res) => {
  try {
    const tenantId = req.user.tenant_id;
    const notification = await db('notifications')
      .where({ id: req.params.notificationId, user_id: req.user.id, tenant_id: tenantId })
      .first();
    if (!notification) return res.status(404).json({ detail: 'Notification not found' });

    await db('notifications').where({ id: req.params.notificationId, tenant_id: tenantId }).del();
    res.status(204).send();
  } catch (err) {
    console.error('Delete notification error:', err);
    res.status(500).json({ detail: 'Internal server error' });
  }
});

export default router;
